eitprocessing/binreader/sequence.py

```diff
@@ -11,141 +11,206 @@
 import itertools
 from dataclasses import dataclass
 from dataclasses import field
+from enum import auto
 from pathlib import Path
 from typing import Dict
 from typing import List
 from typing import Tuple
 import numpy as np
+from strenum import LowercaseStrEnum
 from .event import Event
 from .frameset import Frameset
 from .phases import MaxValue
 from .phases import MinValue
 from .phases import PhaseIndicator
+from .phases import QRSMark
 from .reader import Reader
 from .timing_error import TimingError
 
 
-@dataclass
+class Vendor(LowercaseStrEnum):
+    DRAEGER = auto()
+    TIMPEL = auto()
+    SENTEC = auto()
+    DRAGER = DRAEGER
+    DRÄGER = DRAEGER
+
+
+@dataclass(eq=False)
 class Sequence:
-    path: Path
+    path: Path | str = None
     time: np.ndarray = None
     n_frames: int = None
-    framerate: int = 20
+    framerate: int = None
     framesets: Dict[str, Frameset] = field(default_factory=dict)
     events: List[Event] = field(default_factory=list, repr=False)
     timing_errors: List[TimingError] = field(default_factory=list, repr=False)
     phases: List[PhaseIndicator] = field(default_factory=list, repr=False)
+    vendor: Vendor = None
 
-    def __len__(self):
+    def __len__(self) -> int:
         return self.n_frames
 
-    @classmethod
-    def merge(cls, a, b):  # pylint: disable = too-many-locals
+    def __eq__(self, other) -> bool:
+        for attr in ["n_frames", "framerate", "framesets", "vendor"]:
+            if getattr(self, attr) != getattr(other, attr):
+                return False
+
+        for attr in ["time", "phases", "events", "timing_errors"]:
+            self_attr, other_attr = getattr(self, attr), getattr(other, attr)
+            if len(self_attr) != len(other_attr):
+                return False
+            if not np.all(np.equal(self_attr, other_attr)):
+                return False
+
+        return True
 
+    @classmethod
+    def merge(cls, a, b) -> "Sequence":  # pylint: disable = too-many-locals
```

Comment:
Can you explain to me why `"Sequence"` is/needs to be between `"`s? I don't know this semantic but I notice that if I remove the `"`s, then VSCode doesn't recognize the type hint any more.

Reply:
That is because Sequence hasn't been defined yet when this line is parsed. Only after the entire definition of Sequence is done you can actually refer to it, but methods internal to Sequence can't refer to itself. It's called forward referencing (https://peps.python.org/pep-0484/#forward-references). In Python 3.11 this (and some other issues) is solved with Self as a self-reference (https://peps.python.org/pep-0673/).

Reply:
Cool! Good to know.